File: Java/FractionPerformance.java
import java.util.Arrays;
import java.util.Random;

class FractionPerformance
{
  public class Statistics {
    public double average;
    public double standard_deviation;
    public int sample_size;
    public int median;
    public int mode;

    public Fraction Average() { Fraction f=new Fraction(average); f.Round(100); return f; }
    public Fraction StandardDeviation()  { Fraction f=new Fraction(standard_deviation); f.Round(100); return f; }
    public int Size() { return sample_size; }
    public int Median() { return median; }
    public int Mode() { return mode; }
  }

  public class Frequency implements Comparable<Frequency>
  {
    public int value;
    public int frequency;

    public Frequency() { value=0; frequency=0; }
    public int compareTo(Frequency anotherFrequency)
    {
      return value - anotherFrequency.value;
    }
  }

  public class FrequencyArray
  {
    Frequency [] frequencies;
    private int capacity;
    private int size;
    private int maxFreq;
    public FrequencyArray()
    {
      capacity=16; size=0; maxFreq=0; frequencies=new Frequency[capacity];
    }

    public void Increment(int value)
    {
      int i;
      for(i=0;i<size;i++) {
        if(frequencies[i].value == value)
          break;
      }
      if(i== size) {
        if(i==capacity) {
          Frequency [] temp=frequencies;
          capacity <<=1;
          frequencies=new Frequency[capacity];
          System.arraycopy(temp,0,frequencies,0,size);
        }
        size++;
        frequencies[i] = new Frequency();
        frequencies[i].value=value;
        frequencies[i].frequency=0;
      }
      frequencies[i].frequency++;
    }

    public void Sort()
    {
      Arrays.sort(frequencies,0,size);
    }

    Statistics Statistics()
    {
      Statistics s = new Statistics();

      // Calculate sample size, sum, mode, and average
      int i,sum=0;
      s.sample_size=0;
      maxFreq=0;
      for(i=0;i<size;i++) {
        s.sample_size+=frequencies[i].frequency;
        sum+=frequencies[i].frequency*frequencies[i].value;
        if(frequencies[i].frequency>maxFreq) {
          maxFreq=frequencies[i].frequency;
          s.mode=frequencies[i].value;
        }
      }
      s.average = (double)sum/(double)s.sample_size;

      // Median and variance
      double var=0;
      int count=0;
      s.median=-1;
      for(i=0;i<size;i++) {
        var=(frequencies[i].value-s.average)*(frequencies[i].value-s.average)*frequencies[i].frequency;
        if(s.median == -1) {
          count += frequencies[i].frequency;
          if(count >= s.sample_size/2)
            s.median = frequencies[i].value;
        }
      }

      // Standard deviation
      s.standard_deviation=Math.sqrt(var/(double)(s.sample_size-1));

      return s;
    }

    void DisplayGraph(String xlabel,String ylabel)
    {
      System.out.println("\n"+xlabel+"              "+ylabel);
      System.out.println("-----------------------------------------------------------------");
      double scale=50.0/maxFreq;
      int i;
      for(i=0;i<size;i++) {
        int height=(int)((double)frequencies[i].frequency*scale);
        System.out.println(frequencies[i].value+"|"+"#".repeat(height)+" "+frequencies[i].frequency);
      }
      System.out.println();
    }


    public void ShowResults(String xlabel)
    {
      System.out.println("Max "+xlabel+": "+frequencies[size-1].value);
      Statistics s = Statistics();
      System.out.println("Sample size: "+s.sample_size);
      System.out.println("Average: "+s.Average().toMixedString());
      System.out.println("Median: "+s.Median());
      System.out.println("Mode: "+s.Mode());
      System.out.println("Standard deviation: "+s.StandardDeviation().toMixedString());
      DisplayGraph(xlabel,"Frequency");
    }

    public int Size() { return size; }
  }

  void DoTest(int denominator,FrequencyArray tick_freq,FrequencyArray loop_freq)
  {
    Fraction f = new Fraction();
    for(int i=0;i<denominator;i++) {
      long start = System.nanoTime();;
      f.set((double)i/(double)denominator);
      long elapsed = System.nanoTime() - start;
      if(i>0) {  // First one can skew results
        tick_freq.Increment((int)(elapsed/1000));
        loop_freq.Increment(Fraction.loops);
      }
    }
  }

  void SingleTest(int denominator)
  {
    FrequencyArray tick_freq = new FrequencyArray();
    FrequencyArray loop_freq = new FrequencyArray();
    DoTest(denominator,tick_freq,loop_freq);
    tick_freq.Sort();
    tick_freq.ShowResults("Ticks");
    if(loop_freq.Size() > 1) {
      loop_freq.Sort();
      loop_freq.ShowResults("Loops");
    } else {
      System.out.println("\nStatistics for loop count not gathered. To enable loop statistics,");
      System.out.println("edit Fraction.java and set \"calculate_loop_statistics\" to true.");
    }
  }

  void RandomTest(int minTests)
  {
    int denominators[]=new int[100];
    int nDenominators=0;
    int nTests=0;
    Random ran = new Random();
    while(nTests < minTests && nDenominators < 100) {
      int denominator=ran.nextInt(minTests) + 1000;
      // avoid duplicate denominators
      boolean found=false;
      for(int i=0;i<nDenominators && !found;i++) {
        if(denominator == denominators[i])
          found=true;
      }
      if(!found) {
        denominators[nDenominators++]=denominator;
        nTests += denominator-1;
      }
    }

    FrequencyArray tick_freq = new FrequencyArray();
    FrequencyArray loop_freq = new FrequencyArray();
    for(int i=0;i<nDenominators;i++)
      DoTest(denominators[i],tick_freq,loop_freq);
    tick_freq.Sort();
    tick_freq.ShowResults("Ticks");
    if(loop_freq.Size() > 1) {
      loop_freq.Sort();
      loop_freq.ShowResults("Loops");
    } else {
      System.out.println("\nStatistics for loop count not gathered. To enable loop statistics,");
      System.out.println("edit Fraction.java and set \"calculate_loop_statistics\" to true");
    }
  }

  static void Syntax(String pgm)
  {
//    string pgm = System.Process.GetCurrentProcess().ProcessName;
    System.out.format("Syntax: %s [-h | --help ]\n",pgm);
    System.out.format("        %s [ [-s | --single] N] [ [-r | --random] N ]\n",pgm);
    System.out.format("        %s\n\n",pgm);
    System.out.format("Where:  -h | --help prints this help message\n");
    System.out.format("        -s | --single N -- gather statistics using N as denominator (runs tests using fractions 1/N to (N-1)/N)\n");
    System.out.format("        -r | --random N -- gather statistics running a minimum of N tests using random denominators\n");
    System.out.format("        The default is to run a single test using 1000 as denominator and 1000 minimum random tests\n\n");
    System.out.format("Examples\n");
    System.out.format("   1) To run default case\n");
    System.out.format("      %s\n\n",pgm);
    System.out.format("   2) To run single test using denominator of 100000\n");
    System.out.format("      %s -s 100000\n\n",pgm);
    System.out.format("   3) To run a minimum of 30000 random test\n");
    System.out.format("      %s -r 30000\n\n",pgm);
    System.out.format("   4) To run a single test using denominator of 100000 and a minimum of 30000 random test\n");
    System.out.format("      %s --single 100000 --random 30000\n\n",pgm);
  }

  public static void main(String[] args) {
    FractionPerformance fp= new FractionPerformance();
    if(args.length > 0) {
      int minTest=-1;
      int denominator=-1;
      for(int i=0;i<args.length;i++) {
        if(args[i].equals("-h") || args[i].equals("--help")) {
          Syntax("FractionPerformance");
          return;
        } else if(args[i].equals("-s") || args[i].equals("--single")) {
          if(i == (args.length-1)) {
            System.out.format("Value needs to be supplied for \"%s\" option. ",args[i]);
            Syntax("FractionPerformance");
            return;
          }
          i++;
          try {
            denominator=Integer.parseInt(args[i]);
          } catch(NumberFormatException ne1) {
            System.out.format("Invalid value \"%s\" for \"%s\" option. ",args[i],args[i-1]);
            Syntax("FractionPerformance");
            return;
          }
        } else if(args[i].equals("-r") || args[i].equals("--random")) {
          if(i == (args.length-1)) {
            System.out.format("Value needs to be supplied for \"%s\" option. ",args[i]);
            Syntax("FractionPerformance");
            return;
          }
          i++;
          try {
            minTest=Integer.parseInt(args[i]);
          } catch(NumberFormatException ne2) {
            System.out.format("Invalid value \"%s\" for \"%s\" option. ",args[i],args[i-1]);
            Syntax("FractionPerformance");
            return;
          }
        } else {
          System.out.format("Invalid options specified %s\n",args[i]);
          Syntax("FractionPerformance");
          return;
        }
      }
      if(denominator > 0)
        fp.SingleTest(denominator);
      if(minTest > 0)
        fp.RandomTest(minTest);
    } else {
      fp.SingleTest(1000);
      fp.RandomTest(1000);
    }
  }
}

File: Go/fraction_performance.go
package main

import ("fmt"; "time";. "fraction"; "math"; "strings";"sort"; "math/rand"; "flag" ; "os")

type Statistics struct {
  average float64
  standard_deviation float64
  size int
  median int
  mode int
  maxFreq int
}

func(s Statistics) Average() Fraction {
  var avg Fraction
  avg.Set(s.average)
  avg = avg.Round(100)
  return avg
}

func(s Statistics) StandardDeviation() Fraction {
  var sd Fraction
  sd.Set(s.standard_deviation)
  sd = sd.Round(100)
  return sd
}

type Frequency struct {
  value int
  frequency int
}

type FrequencyArray [] Frequency

/* Sorting interface functions */
func(fa FrequencyArray) Len() int { return len(fa) }
func(fa FrequencyArray) Less(i,j int) bool { return fa[i].value < fa[j].value}
func(fa FrequencyArray) Swap(i,j int) { fa[i],fa[j] = fa[j],fa[i]}

func(fa *FrequencyArray) Increment(value int) {
  var i int
  for i = 0; i<fa.Len(); i++ {
    if (*fa)[i].value == value {
      break
    }
  }
  if i == fa.Len() {
    *fa = append(*fa,Frequency {value,0})
    (*fa)[i].value=value
  }
  (*fa)[i].frequency++
}

func(fa *FrequencyArray) Statistics() Statistics {
  var s Statistics

  // Calculate sample size, sum, mode, and average
  var i,sum int
  sum = 0
  s.maxFreq =0
  s.size=0
  for i=0; i < fa.Len(); i++ {
    s.size+=(*fa)[i].frequency
    sum+=(*fa)[i].frequency*(*fa)[i].value
    if((*fa)[i].frequency > s.maxFreq) {
      s.maxFreq=(*fa)[i].frequency
      s.mode=(*fa)[i].value
    }
  }
  s.average = float64(sum)/float64(s.size)

  // Get median and variance
  var variance float64 = 0.0
  var count int
  s.median=-1
  for i=0; i<fa.Len(); i++ {
    variance+=(float64((*fa)[i].value) - s.average)*(float64((*fa)[i].value) - s.average)*float64((*fa)[i].frequency)
    if s.median == -1 {
      count+=(*fa)[i].frequency
      if count >= s.size/2 {
        s.median=(*fa)[i].value
      }
    }
  }

  // standard deviation
  s.standard_deviation = math.Sqrt(variance/float64(s.size-1))
  return s
}

var terminalCols int = 50
func(fa *FrequencyArray) DisplayGraph(xlabel string,ylabel string,maxValue int) {
  fmt.Printf("\n%v|                   %v\n",xlabel,ylabel)
  fmt.Println(strings.Repeat("-",terminalCols+10))
  var scale float64 = float64(terminalCols)/float64(maxValue)
  var i int
  for i=0; i < fa.Len(); i++ {
    var height int = int(math.Round(scale*float64((*fa)[i].frequency)))
//    fmt.Println((*fa)[i].value,"|",strings.Repeat("#",height)," ",(*fa)[i].frequency)
    fmt.Printf("%4d |%s %d\n",(*fa)[i].value,strings.Repeat("#",height),(*fa)[i].frequency)
  }
}

func (fa *FrequencyArray) ShowResults(xlabel string) {
  fmt.Println("Max ",xlabel,": ",(*fa)[fa.Len()-1].value)
  s := fa.Statistics();
  fmt.Println("Sample size: ",s.size)
  fmt.Println("Average: ",s.Average().MixedString())
  fmt.Println("Median: ",s.median)
  fmt.Println("Mode: ",s.mode)
  fmt.Println("Standard Deviation: ",s.StandardDeviation().MixedString())
  fa.DisplayGraph(xlabel,"Frequency",s.maxFreq)
}

func DoTest(denominator int,time_freq *FrequencyArray,loop_freq *FrequencyArray) {
  var i int
  var f Fraction
  for i=0; i < denominator; i++ {
    var value float64 = float64(i)/float64(denominator)
    start := time.Now()
    f.Set(value)
    elapsed := time.Now().Sub(start)
    if(i>0) {
      time_freq.Increment(int(math.Round(float64(elapsed.Nanoseconds())/100.0)))
      loop_freq.Increment(Loops)
    }
  }
}

func SingleTest(denominator int) {
  var time_freq FrequencyArray
  var loop_freq FrequencyArray
  DoTest(denominator,&time_freq,&loop_freq)
  sort.Sort(FrequencyArray(time_freq))
  sort.Sort(FrequencyArray(loop_freq))
  time_freq.ShowResults("t(us)")
  fmt.Println("")
  loop_freq.ShowResults("Loops")
}

func RandomTest(minTests int) {
  rand.Seed(time.Now().UnixNano())
  var denominators [] int
  var nTests int = 0
  for nTests < minTests {
    var denominator int = rand.Intn(minTests+100)
    var found bool = false
    for _,d := range denominators {
      if(d == denominator) {
        found=true
        break
      }
    }
    if !found {
      denominators=append(denominators,denominator)
      nTests += denominator-1
    }
  }

  var time_freq FrequencyArray
  var loop_freq FrequencyArray
  for _,denominator := range denominators {
    DoTest(denominator,&time_freq,&loop_freq)
  }
  sort.Sort(FrequencyArray(time_freq))
  sort.Sort(FrequencyArray(loop_freq))
  time_freq.ShowResults("t(us)")
  fmt.Println("")
  loop_freq.ShowResults("Loops")
}

func Syntax() {
  fmt.Println("Syntax: ",os.Args[0]," -h");
  fmt.Println("        ",os.Args[0]," -s N");
  fmt.Println("        ",os.Args[0]," -r N");
  fmt.Println("        ",os.Args[0]," -s N -r N");
  fmt.Println("        ",os.Args[0],"\n");
  fmt.Println("Where:  -h prints this help message");
  fmt.Println("        -s N -- gather statistics using N as denominator (runs tests using fractions 1/N to (N-1)/N)");
  fmt.Println("        -r N -- gather statistics running a minimum of N tests using random denominators");
  fmt.Println("        No options -- Run the default case of a single test using 1000 as denominator and 1000 minimum random tests\n");
  fmt.Println("Examples");
  fmt.Println("   1) To run default case");
  fmt.Println("      ",os.Args[0],"\n");
  fmt.Println("   2) To run single test using denominator of 100000");
  fmt.Println("      ",os.Args[0]," -s 100000\n");
  fmt.Println("   3) To run a minimum of 30000 random test");
  fmt.Println("      ",os.Args[0]," -r 30000\n");
  fmt.Println("   4) To run a single test using denominator of 100000 and a minimum of 30000 random test");
  fmt.Println("      ",os.Args[0]," -s 100000 -r 30000\n");
}

func main() {
  args := os.Args
  if len(args) > 1 {
    fs := flag.NewFlagSet("fraction_performance", flag.ExitOnError)
    fs.Usage = Syntax
//    h := fs.Bool("h", false, "Shows detailed help.")
    var singleDenominator int = 0
    var minTest int = 0
    fs.IntVar(&singleDenominator,"s",0,"Run single test using denominator specified")
    fs.IntVar(&minTest,"r",0,"Run test using random denominators for a minimum of the test specified")
    if err := fs.Parse(os.Args[1:]); err != nil {
                os.Exit(100)
        }
        //flag.Parse()
    args := fs.Args()
    if len(args) > 0 {
      fmt.Println("\n****Invalid arguments specified***")
      Syntax()
      os.Exit(100)
    }
    if singleDenominator > 0 {
      SingleTest(singleDenominator)
    }
    if minTest > 0 {
      RandomTest(minTest)
    }
  } else {
    SingleTest(1000)
    RandomTest(1000)
  }
}
